### PR TITLE
 Add optional header support for TSV input files

### PR DESCRIPTION
**Pull request by Bibisha06**
Resolves #33.

This PR adds support for an optional header row that, when present, is automatically detected and used to drive column mapping — allowing columns in any order, missing optional columns, and extra columns to all be handled correctly.

## Changes

- `lms/util/tsv.py` — new shared utility with `read_tsv()` and `_parse_header()`. Automatically detects a header if any cell in the first row matches a known column name (case-insensitive). Falls back to positional ordering if no header is found.
- `lms/cli/courses/assignments/scores/upload.py` — refactored `_load_scores` to use `read_tsv` with `['user', 'score', 'comment']` as expected columns.
- `lms/cli/courses/groupsets/memberships/common.py` — refactored `load_group_memberships` to use `read_tsv` with `['group', 'user']` as expected columns.

All existing behavior and error messages are preserved unchanged.

## Corner Cases Covered

|                              | In Order | Out of Order |
|------------------------------|----------|--------------|
| All columns                  | ✅       | ✅           |
| Missing columns              | ✅       | ✅           |
| Extra columns                | ✅       | ✅           |
| No header (positional fallback) | ✅    | —            |

## Test Plan

New test fixtures added for each of the 6 header corner cases under `lms/backend/testdata/`. All existing tests pass.

### CODE REVIEW

**Review by eriq-augustine — CHANGES_REQUESTED** [review]
There are many style issues here that need to be addressed before we start any review.
The number one rule of style is to be consistent with the code around you.
Just take look at any of the other code in this project.

File: lms/util/tsv.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+"""
+Utilities for parsing TSV files.
+"""
+import typing
+import edq.util.dirent
+
+def read_tsv(path: str,
```
Comment (eriq-augustine) [inline lms/util/tsv.py]: These should probably be one or more override options to prevent interpretation.
Like if a user did not want us to try and guess headers.

(This may be obviated by having the multiple functions discussed in another comment.)

**Review by eriq-augustine — CHANGES_REQUESTED** [review]
A good start.

File: lms/util/tsv.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    with open(path, 'r', encoding = edq.util.dirent.DEFAULT_ENCODING) as file:
+        lineno = 0
+        real_rows = 0
+        header_map: typing.Optional[typing.Dict[str, int]] = None
+        found_header = False
+
+        for line in file:
+            lineno += 1
+            line = line.rstrip('\r\n')
```
[the remark this answers is not in the document]
Comment (eriq-augustine) [reply lms/util/tsv.py]: But there is no difference in a row that is short some columns and a row with empty values in its final columns.

File: lms/util/tsv.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
+"""
+Utilities for parsing TSV files.
```
Comment (eriq-augustine) [inline lms/util/tsv.py]: I don't think this comment is necessary.
It is also inaccurate, because there is only really one utility in this file.

**Review by eriq-augustine — CHANGES_REQUESTED** [review]
!

Your code does not pass CI.
Make sure to run checks locally and in your fork.

You still have a lot of style issues.
When you are doing things, check other places in the codebase to see if anyone else does it that way.
Generally, you want to make your code the same as the code around it.
On subsequent reviews, I am going to stop earlier if I see too much of code that does not follow our codebase's patterns.

Try to be more aware of your vertical whitespace usage.
You are missing whitespace when you need to separate vertical blocks,
and have extra whitespace when it is not necessary.

File: lms/backend/testdata/cli/data/scores/course101_hw0_emptyvalue.txt [hunk 4]
```diff
@@ -1,2 +1,2 @@
 User	Score
-course-student@test.edulinq.org	
+course-student@test.edulinq.org
```
Comment (eriq-augustine) [inline lms/backend/testdata/cli/data/scores/course101_hw0_emptyvalue.txt:2]: Unless I am missing something, it seems like you changed the case that this test is testing.
Shouldn't there be one case with a missing value (no tab) and one with an empty value (tab)?

File: lms/util/tsv.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        self.parts = parts
+        self.lineno = lineno
+        self.effective_lineno = effective_lineno
+
+    def raise_error(self, message: str) -> typing.NoReturn:
+        """ Raise a ValueError with this row's line context prepended. """
+
+        raise ValueError(
+            f"Line {self.lineno}"
+            f" (effective {self.effective_lineno}): {message}"
```
Comment (eriq-augustine) [inline lms/util/tsv.py:27]: "effective" doesn't mean anything to the user.
Try to make this a bit more clear.

File: lms/backend/testdata/cli/tests/courses/assignments/scores/upload_course101_hw0_longline.txt [hunk 6]
```diff
@@ -9,4 +9,4 @@
     "error": true,
 }
 ---
-builtins.ValueError: File '__DATA_DIR__(scores/course101_hw0_longline.txt)' line 2 has the incorrect number of values. Expecting 2-3, found 4.
+builtins.ValueError: Line 2 (effective 2): Has the incorrect number of values. Expecting 2-3, found 4.
```
Comment (eriq-augustine) [inline lms/backend/testdata/cli/tests/courses/assignments/scores/upload_course101_hw0_longline.txt:12]: You removed the path here, which makes it much harder for the user to figure out what is going on.

File: lms/cli/courses/assignments/scores/upload.py [hunk 7]
```diff
@@ -43,49 +42,60 @@ def run_cli(args: argparse.Namespace) -> int:
 def _load_scores(
         backend: lms.model.backend.APIBackend,
         path: str,
-        skip_rows: bool,
+        skip_rows: int,
```
Comment (eriq-augustine) [inline lms/cli/courses/assignments/scores/upload.py:45]: Here you changed the semantics, but left the old name.
This makes it far too easy for devs to use the new functionality improperly.
Change the param's name.

File: lms/cli/courses/assignments/scores/upload.py [hunk 8]
```diff
@@ -43,49 +42,60 @@ def run_cli(args: argparse.Namespace) -> int:
...
-            parts = [part.strip() for part in line.split("\t")]
-            if (len(parts) not in [2, 3]):
-                raise ValueError(f"File '{path}' line {lineno} has the incorrect number of values. Expecting 2-3, found {len(parts)}.")
+    for row in tsv.rows:
+        if (tsv.headers is None):
+            if (len(row.parts) not in [2, 3]):
+                row.raise_error(f"Has the incorrect number of values. Expecting 2-3, found {len(row.parts)}.")
+        else:
+            if (len(row.parts) <= max_required_index):
+                row.raise_error(f"Has the incorrect number of values. Expecting 2-3, found {len(row.parts)}.")
```
Comment (eriq-augustine) [inline lms/cli/courses/assignments/scores/upload.py:73]: This seems pretty clunky.
Seems like something that could have been checked by the common functionality.

File: lms/util/tsv.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+    def __init__(
+            self,
+            parts: typing.List[str],
+            lineno: int,
+            effective_lineno: int,
+            ) -> None:
+        self.parts = parts
+        self.lineno = lineno
+        self.effective_lineno = effective_lineno
```
Comment (eriq-augustine) [inline lms/util/tsv.py:20]: Missing docstrings.

File: lms/util/tsv.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+        rows.append(TSVRow(
+            parts = parts,
+            lineno = lineno,
+            effective_lineno = effective_lineno
+        ))
+
+    return rows
+
+def load_tsv_table(
```
Comment (eriq-augustine) [inline lms/util/tsv.py:88]: This name is at odds with your type name.
This function name says it returns a "table",
but your type says it is a "file".

File: lms/util/tsv.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    if (index is None):
+        return ''
+
+    if (index < len(row.parts)):
+        return row.parts[index]
+
+    return ''
+
+def parse_tsv_rows(path: str) -> typing.List[TSVRow]:
+    """ Parse a TSV file into raw rows, skipping blanks. Parts are stripped. """
```
Comment (eriq-augustine) [inline lms/util/tsv.py:67]: Break comment lines on sentences.
This makes diffs easier.

File: lms/util/tsv.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+
+    def raise_error(self, message: str) -> typing.NoReturn:
+        """ Raise a ValueError with this row's line context prepended. """
+
+        raise ValueError(
+            f"Line {self.lineno}"
+            f" (effective {self.effective_lineno}): {message}"
+        )
+
+class TSVFile:
```
Comment (eriq-augustine) [inline lms/util/tsv.py:30]: This type does not seem particularly useful to me.

I see two different use cases for callers of this module:
 - They want to parse a raw TSV file and do all the interpretation themselves.
   -  Currently solved by parse_tsv_rows().
 - They have expected headers/columns and want the module to do their error handling and get back a list of dicts keyed with the columns.

This is kind of in-between, but doesn't seem like what people will want.

File: lms/util/tsv.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    if (len(expected_columns) == 0):
+        raise ValueError("expected_columns must not be empty.")
+
+    if (skip_rows < 0):
+        raise ValueError("skip_rows must be non-negative.")
+
+    if (required_columns is not None):
+        expected_lower = {col.lower() for col in expected_columns}
+        unknown = []
+        for c in required_columns:
```
Comment (eriq-augustine) [inline lms/util/tsv.py:106]: Be more expressive with your variable names.
(There are only a few exceptions.)

File: lms/util/tsv.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+            if (c.lower() not in expected_lower):
+                unknown.append(c)
+
+        if (len(unknown) > 0):
+            raise ValueError(
+                f"required_columns contains names not in expected_columns:"
+                f" {sorted(unknown)}"
+            )
```
Comment (eriq-augustine) [inline lms/util/tsv.py:114]: It seems like you can resolve this potential mismatch by changing your inputs instead.

For example, instead of both `expected_columns` and `required_columns`,
what about just `columns` that takes a str (column name) or tuple(str, bool) (name and required)?
This would make it much harder for the caller to mess up
(this makes your core more defensive and friendly).

File: lms/util/tsv.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+def _match_header_row(
+        parts: typing.List[str],
+        expected_columns: typing.List[str],
+        required_columns: typing.List[str],
+        ) -> typing.Union[typing.Dict[str, typing.Union[int, None]], None]:
+    """ Return a column index map if all required columns match, else None. """
+
+    normalized_expected = {col.lower(): col for col in expected_columns}
+    normalized_required = {col.lower() for col in required_columns}
```
Comment (eriq-augustine) [inline lms/util/tsv.py:153]: You are defining a semantic that you have not told the user about anywhere.
You are deciding that all column names are case insensitive (which I do not want).

File: lms/util/tsv.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+            raise ValueError(
+                f"required_columns contains names not in expected_columns:"
+                f" {sorted(unknown)}"
+            )
+
+    raw_rows = parse_tsv_rows(path)
+    data_rows = raw_rows[skip_rows:]
+
+    default_map: typing.Dict[str, typing.Union[int, None]] = {}
+    for i, name in enumerate(expected_columns):
```
Comment (eriq-augustine) [inline lms/util/tsv.py:120]: We generally put parents around our iteration tuples.

File: lms/util/tsv_test.py [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+def _write_temp(content: str) -> str:
+    """ Write content to a temp file and return its path. """
+
+    path = edq.util.dirent.get_temp_path(suffix = '.tsv')
+    with open(path, 'w', encoding = edq.util.dirent.DEFAULT_ENCODING) as f:
+        f.write(content)
+
+    return path
+
+def _remove_temp(path: str) -> None:
```
Comment (eriq-augustine) [inline lms/util/tsv_test.py:15]: Why?
Our temp files already remove themselves.

File: lms/util/tsv_test.py [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+    def test_tracks_real_and_effective_line_numbers(self):
+        """ lineno counts blank lines; effective_lineno does not. """
+
+        path = _write_temp("foo\tbar\n\nbaz\tqux\n")
+        try:
+            rows = lms.util.tsv.parse_tsv_rows(path)
+            self.assertEqual(rows[0].lineno, 1)
+            self.assertEqual(rows[0].effective_lineno, 1)
+            self.assertEqual(rows[1].lineno, 3)
+            self.assertEqual(rows[1].effective_lineno, 2)
```
Comment (eriq-augustine) [inline lms/util/tsv_test.py:64]: Just make what you expect the output to be and assert against that.

File: lms/util/tsv_test.py [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+import edq.testing.unittest
+
+import lms.util.tsv
+
+class TestTsv(edq.testing.unittest.BaseTest):
+    """
+    Test TSV parsing.
+    """
+
+    def test_read_tsv_with_header(self):
```
[the remark this answers is not in the document]
Comment (eriq-augustine) [reply lms/util/tsv_test.py]: !

This comment has not been addressed.
You have like a dozen tests that could just be test cases.

**Review by eriq-augustine — CHANGES_REQUESTED** [review]
I ran out of review time on the tests,
so my comments are incomplete.
But there is plenty to work on.

There are a good amount of design issues.
Make sure to think about what things look like for the caller of your code.

Your code [smells a bit](https://en.wikipedia.org/wiki/Code_smell) here.
Notice that your code that invokes the common functionality
(e.g., in lms/cli/courses/assignments/scores/upload.py)
is about the same number of lines for no new functionality.
This is only a smell (since that can sometimes make sense),
but it feels a bit off here.

**Comment by eriq-augustine** [comment]
Based on the inactivity and user's change to draft,
I am going to close this PR.
Feel free to make a new one if you want to try again.
